cartographer_ros/cartographer_ros/map_builder_bridge.cc

```diff
@@ -175,4 +175,124 @@ SensorBridge* MapBuilderBridge::sensor_bridge(const int trajectory_id) {
   return sensor_bridges_.at(trajectory_id).get();
 }
 
+visualization_msgs::MarkerArray MapBuilderBridge::GetTrajectoryNodesList() {
+  visualization_msgs::MarkerArray trajectory_nodes_list;
+
+  for (int trajectory_id = 0;
+       trajectory_id < map_builder_.num_trajectory_builders();
+       ++trajectory_id) {
+    const auto trajectory_nodes =
+        map_builder_.sparse_pose_graph()->GetTrajectoryNodes();
+
+    int i = 0;
+    for (const auto& node : trajectory_nodes) {
+      visualization_msgs::Marker marker;
+      marker.id = i++;
+      marker.type = visualization_msgs::Marker::ARROW;
+      marker.header.stamp = ::ros::Time::now();
+      marker.header.frame_id = options_.map_frame;
+      marker.color.b = 1.0;
+      marker.color.a = 1.0;
+      marker.scale.x = 0.1;
+      marker.scale.y = 0.05;
+      marker.scale.z = 0.05;
+      marker.pose = ToGeometryMsgPose(node.pose);
+      trajectory_nodes_list.markers.push_back(marker);
+    }
+  }
+  return trajectory_nodes_list;
+}
+
+std::unordered_map<std::string, visualization_msgs::MarkerArray>
+MapBuilderBridge::GetConstraintsList() {
+  std::unordered_map<std::string, visualization_msgs::MarkerArray>
+      constratins_viz;
+
+  const auto trajectory_nodes =
+      map_builder_.sparse_pose_graph()->GetTrajectoryNodes();
+  const auto constraints = map_builder_.sparse_pose_graph()->constraints();
+  std::vector<cartographer::transform::Rigid3d> submap_transforms;
+  for (auto& trajectory : trajectory_nodes) {
```

Comment:
I'm not sure what this should compute. It is iterating over all nodes and adding for each node the poses of all submaps of the same trajectory, i.e. multiply times depending on how many nodes a trajectory has.